File: software/android/app/src/main/java/app/tucklet/protocol/Protocol.kt
// Protocol.kt
// Kotlin mirror of `tucklet-proto`. Wire-compatible with the firmware: snake_case
// field names and the same shapes (incl. the flattened ItemState).
//
// License: PolyForm Noncommercial 1.0.0
package app.tucklet.protocol

import kotlinx.serialization.*
import kotlinx.serialization.descriptors.*
import kotlinx.serialization.encoding.*
import kotlinx.serialization.json.*

const val PROTOCOL_VERSION: Int = 1
typealias EpochSeconds = Long

@Serializable
enum class RadioKind {
    @SerialName("single_c5") SINGLE_C5,
    @SerialName("dual_c5") DUAL_C5,
}

@Serializable
enum class DataTransport {
    @SerialName("soft_ap") SOFT_AP,
    @SerialName("wifi_aware") WIFI_AWARE,
    @SerialName("wired_usb_hs") WIRED_USB_HS,
}

@Serializable
enum class Platform {
    @SerialName("ios") IOS,
    @SerialName("android") ANDROID,
    @SerialName("desktop") DESKTOP,
}

/** StorageKind matches serde's default enum encoding: the unit variant is the
 *  string "micro_sd"; the struct variant is {"emmc":{"capacity_gib":N}}. */
@Serializable(with = StorageKindSerializer::class)
sealed class StorageKind {
    object MicroSd : StorageKind()
    data class Emmc(val capacityGib: Int) : StorageKind()

    val label: String
        get() = when (this) {
            is MicroSd -> "microSD"
            is Emmc -> "$capacityGib GB built-in"
        }
}

object StorageKindSerializer : KSerializer<StorageKind> {
    override val descriptor: SerialDescriptor =
        buildClassSerialDescriptor("StorageKind")

    override fun serialize(encoder: Encoder, value: StorageKind) {
        val json = (encoder as JsonEncoder)
        when (value) {
            is StorageKind.MicroSd -> json.encodeJsonElement(JsonPrimitive("micro_sd"))
            is StorageKind.Emmc -> json.encodeJsonElement(
                buildJsonObject {
                    putJsonObject("emmc") { put("capacity_gib", value.capacityGib) }
                }
            )
        }
    }

    override fun deserialize(decoder: Decoder): StorageKind {
        val el = (decoder as JsonDecoder).decodeJsonElement()
        return when (el) {
            is JsonPrimitive -> if (el.content == "micro_sd") StorageKind.MicroSd
                else error("unknown storage ${el.content}")
            is JsonObject -> {
                val cap = el["emmc"]?.jsonObject?.get("capacity_gib")?.jsonPrimitive?.int
                    ?: error("emmc missing capacity_gib")
                StorageKind.Emmc(cap)
            }
            else -> error("unexpected storage encoding")
        }
    }
}

@Serializable
data class DeviceCapabilities(
    @SerialName("protocol_version") val protocolVersion: Int,
    @SerialName("firmware_version") val firmwareVersion: String,
    val radio: RadioKind,
    val storage: StorageKind,
    val transports: List<DataTransport>,
    @SerialName("device_id") val deviceId: String,
) {
    fun supports(t: DataTransport) = transports.contains(t)
    fun bestWireless(): DataTransport? =
        transports.firstOrNull { it == DataTransport.WIFI_AWARE }
            ?: transports.firstOrNull { it == DataTransport.SOFT_AP }
}

/** Where an item lives, in the exact vocabulary the UI shows. */
sealed class ItemState {
    object OnPhone : ItemState()
    object OnTucklet : ItemState()
    data class Temporary(val expiresAt: EpochSeconds?) : ItemState()

    val label: String
        get() = when (this) {
            is OnPhone -> "On phone"
            is OnTucklet -> "On Tucklet"
            is Temporary -> "Temporary"
        }
}

@Serializable
enum class TemporaryPolicy(val seconds: Long?) {
    @SerialName("one_hour") ONE_HOUR(3_600),
    @SerialName("one_day") ONE_DAY(86_400),
    @SerialName("one_week") ONE_WEEK(604_800),
    @SerialName("keep") KEEP(null);

    val label: String
        get() = when (this) {
            ONE_HOUR -> "1 hour"; ONE_DAY -> "1 day"; ONE_WEEK -> "1 week"; KEEP -> "Keep"
        }
}

@Serializable
data class OriginMetadata(
    val platform: Platform,
    val app: String,
    val collection: String,
    val album: String? = null,
    @SerialName("device_name") val deviceName: String,
)

/** MediaItem with the ItemState FLATTENED to the top level (state + expires_at),
 *  matching the firmware's #[serde(flatten)]. The convenient `itemState` is
 *  derived; the serialized fields are `state`/`expires_at`. */
@Serializable
data class MediaItem(
    val id: String,
    val name: String,
    @SerialName("size_bytes") val sizeBytes: Long,
    val mime: String,
    @SerialName("created_at") val createdAt: EpochSeconds,
    val origin: OriginMetadata,
    val state: String,                              // "on_phone"|"on_tucklet"|"temporary"
    @SerialName("expires_at") val expiresAt: EpochSeconds? = null,
    val checksum: String? = null,
) {
    val isImage get() = mime.startsWith("image/")
    val isVideo get() = mime.startsWith("video/")

    val itemState: ItemState
        get() = when (state) {
            "on_phone" -> ItemState.OnPhone
            "on_tucklet" -> ItemState.OnTucklet
            "temporary" -> ItemState.Temporary(expiresAt)
            else -> ItemState.OnPhone
        }

    companion object {
        fun onPhone(
            id: String, name: String, sizeBytes: Long, mime: String,
            createdAt: EpochSeconds, origin: OriginMetadata,
        ) = MediaItem(id, name, sizeBytes, mime, createdAt, origin, "on_phone", null, null)
    }
}

@Serializable
data class Manifest(
    val items: List<MediaItem>,
    @SerialName("free_bytes") val freeBytes: Long,
    @SerialName("total_bytes") val totalBytes: Long,
)

@Serializable
data class StatusReport(
    @SerialName("battery_percent") val batteryPercent: Int,
    val charging: Boolean,
    @SerialName("free_bytes") val freeBytes: Long,
    @SerialName("total_bytes") val totalBytes: Long,
    @SerialName("card_present") val cardPresent: Boolean,
    @SerialName("firmware_version") val firmwareVersion: String,
)

@Serializable
data class PairRequest(
    @SerialName("phone_pubkey") val phonePubkey: String,
    @SerialName("phone_name") val phoneName: String,
)

@Serializable
data class PairResponse(
    val paired: Boolean,
    @SerialName("device_pubkey") val devicePubkey: String? = null,
    val reason: String? = null,
)

@Serializable
data class SessionRequest(
    @SerialName("challenge_signature") val challengeSignature: String,
    val transport: DataTransport,
)

@Serializable
data class SessionGrant(
    @SerialName("ssid_or_service") val ssidOrService: String,
    val psk: String,
    val ip: String,
    val token: String,
    @SerialName("ttl_seconds") val ttlSeconds: Int,
)

@Serializable
enum class Command {
    @SerialName("sleep") SLEEP,
    @SerialName("factory_reset_confirm") FACTORY_RESET_CONFIRM,
    @SerialName("begin_trickle") BEGIN_TRICKLE,
}

@Serializable
enum class TransferKind { @SerialName("offload") OFFLOAD, @SerialName("load") LOAD }

@Serializable
enum class TransferMode { @SerialName("batch") BATCH, @SerialName("trickle") TRICKLE }

@Serializable
data class TransferItem(
    val id: String,
    @SerialName("size_bytes") val sizeBytes: Long,
    val mime: String,
)

@Serializable
data class TransferProgress(
    @SerialName("items_total") val itemsTotal: Int,
    @SerialName("items_done") val itemsDone: Int,
    @SerialName("bytes_total") val bytesTotal: Long,
    @SerialName("bytes_done") val bytesDone: Long,
    @SerialName("eta_seconds") val etaSeconds: Int,
    @SerialName("throughput_bps") val throughputBps: Long,
)

/** Shared JSON config (lenient like the device). */
val TuckletJson = Json {
    ignoreUnknownKeys = true
    encodeDefaults = false
    explicitNulls = false
}

File: software/ios/Tucklet/Models/Protocol.swift
// Protocol.swift
// Swift mirror of `tucklet-proto` (the Rust source of truth). These Codable
// types decode/encode the exact same JSON the firmware emits, so the app and
// device always agree. Keep this in lockstep with firmware/crates/tucklet-proto.
//
// License: PolyForm Noncommercial 1.0.0

import Foundation

public let protocolVersion: UInt16 = 1
public typealias EpochSeconds = Int64

// MARK: - Variant matrix

public enum RadioKind: String, Codable, Sendable { case singleC5 = "single_c5", dualC5 = "dual_c5" }

public enum StorageKind: Codable, Sendable, Equatable {
    case microSd
    case emmc(capacityGiB: UInt32)

    private enum CodingKeys: String, CodingKey { case microSd = "micro_sd", emmc }
    private struct EmmcPayload: Codable { let capacity_gib: UInt32 }

    public init(from decoder: Decoder) throws {
        // serde encodes externally-tagged enums: {"micro_sd": null} or {"emmc":{"capacity_gib":64}}
        let c = try decoder.container(keyedBy: CodingKeys.self)
        if c.contains(.emmc) {
            let p = try c.decode(EmmcPayload.self, forKey: .emmc)
            self = .emmc(capacityGiB: p.capacity_gib)
        } else {
            self = .microSd
        }
    }
    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .microSd: try c.encodeNil(forKey: .microSd)
        case .emmc(let cap): try c.encode(EmmcPayload(capacity_gib: cap), forKey: .emmc)
        }
    }
}

public enum DataTransport: String, Codable, Sendable, CaseIterable {
    case softAp = "soft_ap"
    case wifiAware = "wifi_aware"
    case wiredUsbHs = "wired_usb_hs"
}

public struct DeviceCapabilities: Codable, Sendable, Equatable {
    public let protocolVersion: UInt16
    public let firmwareVersion: String
    public let radio: RadioKind
    public let storage: StorageKind
    public let transports: [DataTransport]
    public let deviceId: String

    enum CodingKeys: String, CodingKey {
        case protocolVersion = "protocol_version"
        case firmwareVersion = "firmware_version"
        case radio, storage, transports
        case deviceId = "device_id"
    }

    public func supports(_ t: DataTransport) -> Bool { transports.contains(t) }

    /// Best wireless transport, preferring seamless Wi-Fi Aware over SoftAP.
    public var bestWireless: DataTransport? {
        if supports(.wifiAware) { return .wifiAware }
        if supports(.softAp) { return .softAp }
        return nil
    }
}

// MARK: - Item state (plain-language UX vocabulary; never "cache")

public enum ItemState: Codable, Sendable, Equatable {
    case onPhone
    case onTucklet
    case temporary(expiresAt: EpochSeconds?)

    private enum CodingKeys: String, CodingKey { case state, expires_at }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        switch try c.decode(String.self, forKey: .state) {
        case "on_phone": self = .onPhone
        case "on_tucklet": self = .onTucklet
        case "temporary": self = .temporary(expiresAt: try c.decodeIfPresent(EpochSeconds.self, forKey: .expires_at))
        case let other: throw DecodingError.dataCorruptedError(forKey: .state, in: c, debugDescription: "unknown state \(other)")
        }
    }
    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .onPhone: try c.encode("on_phone", forKey: .state)
        case .onTucklet: try c.encode("on_tucklet", forKey: .state)
        case .temporary(let e):
            try c.encode("temporary", forKey: .state)
            try c.encodeIfPresent(e, forKey: .expires_at)
        }
    }

    public var label: String {
        switch self {
        case .onPhone: return "On phone"
        case .onTucklet: return "On Tucklet"
        case .temporary: return "Temporary"
        }
    }
}

public enum TemporaryPolicy: String, Codable, Sendable, CaseIterable {
    case oneHour = "one_hour", oneDay = "one_day", oneWeek = "one_week", keep

    public var lifetimeSeconds: Int64? {
        switch self {
        case .oneHour: return 3_600
        case .oneDay: return 86_400
        case .oneWeek: return 604_800
        case .keep: return nil
        }
    }
    public var label: String {
        switch self {
        case .oneHour: return "1 hour"; case .oneDay: return "1 day"
        case .oneWeek: return "1 week"; case .keep: return "Keep"
        }
    }
    public func resolve(now: EpochSeconds) -> ItemState {
        .temporary(expiresAt: lifetimeSeconds.map { now + $0 })
    }
}

// MARK: - Origin + media

public enum Platform: String, Codable, Sendable { case android, ios, desktop }

public struct OriginMetadata: Codable, Sendable, Equatable {
    public let platform: Platform
    public let app: String
    public let collection: String
    public let album: String?
    public let deviceName: String
    enum CodingKeys: String, CodingKey { case platform, app, collection, album, deviceName = "device_name" }
}

public struct MediaItem: Codable, Sendable, Equatable, Identifiable {
    public let id: String
    public let name: String
    public let sizeBytes: UInt64
    public let mime: String
    public let createdAt: EpochSeconds
    public let origin: OriginMetadata
    public let state: ItemState
    public let checksum: String?

    // `state` is flattened to the top level to match the firmware's
    // #[serde(flatten)] (JSON carries "state" and optional "expires_at"
    // alongside the other fields, not as a nested object).
    enum CodingKeys: String, CodingKey {
        case id, name, sizeBytes = "size_bytes", mime, createdAt = "created_at"
        case origin, checksum
        case state, expires_at
    }

    public init(id: String, name: String, sizeBytes: UInt64, mime: String,
                createdAt: EpochSeconds, origin: OriginMetadata, state: ItemState,
                checksum: String?) {
        self.id = id; self.name = name; self.sizeBytes = sizeBytes; self.mime = mime
        self.createdAt = createdAt; self.origin = origin; self.state = state; self.checksum = checksum
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        name = try c.decode(String.self, forKey: .name)
        sizeBytes = try c.decode(UInt64.self, forKey: .sizeBytes)
        mime = try c.decode(String.self, forKey: .mime)
        createdAt = try c.decode(EpochSeconds.self, forKey: .createdAt)
        origin = try c.decode(OriginMetadata.self, forKey: .origin)
        checksum = try c.decodeIfPresent(String.self, forKey: .checksum)
        // Flattened state, read from the same container.
        switch try c.decode(String.self, forKey: .state) {
        case "on_phone": state = .onPhone
        case "on_tucklet": state = .onTucklet
        case "temporary": state = .temporary(expiresAt: try c.decodeIfPresent(EpochSeconds.self, forKey: .expires_at))
        case let other: throw DecodingError.dataCorruptedError(forKey: .state, in: c, debugDescription: "unknown state \(other)")
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encode(name, forKey: .name)
        try c.encode(sizeBytes, forKey: .sizeBytes)
        try c.encode(mime, forKey: .mime)
        try c.encode(createdAt, forKey: .createdAt)
        try c.encode(origin, forKey: .origin)
        try c.encodeIfPresent(checksum, forKey: .checksum)
        switch state {
        case .onPhone: try c.encode("on_phone", forKey: .state)
        case .onTucklet: try c.encode("on_tucklet", forKey: .state)
        case .temporary(let e):
            try c.encode("temporary", forKey: .state)
            try c.encodeIfPresent(e, forKey: .expires_at)
        }
    }

    public var isImage: Bool { mime.hasPrefix("image/") }
    public var isVideo: Bool { mime.hasPrefix("video/") }
}

public struct Manifest: Codable, Sendable, Equatable {
    public let items: [MediaItem]
    public let freeBytes: UInt64
    public let totalBytes: UInt64
    enum CodingKeys: String, CodingKey { case items, freeBytes = "free_bytes", totalBytes = "total_bytes" }
}

// MARK: - Control plane

public struct StatusReport: Codable, Sendable, Equatable {
    public let batteryPercent: UInt8
    public let charging: Bool
    public let freeBytes: UInt64
    public let totalBytes: UInt64
    public let cardPresent: Bool
    public let firmwareVersion: String
    enum CodingKeys: String, CodingKey {
        case batteryPercent = "battery_percent", charging
        case freeBytes = "free_bytes", totalBytes = "total_bytes"
        case cardPresent = "card_present", firmwareVersion = "firmware_version"
    }
}

public struct PairRequest: Codable, Sendable {
    public let phonePubkey: String
    public let phoneName: String
    enum CodingKeys: String, CodingKey { case phonePubkey = "phone_pubkey", phoneName = "phone_name" }
}

public struct PairResponse: Codable, Sendable {
    public let paired: Bool
    public let devicePubkey: String?
    public let reason: String?
    enum CodingKeys: String, CodingKey { case paired, devicePubkey = "device_pubkey", reason }
}

public struct SessionRequest: Codable, Sendable {
    public let challengeSignature: String
    public let transport: DataTransport
    enum CodingKeys: String, CodingKey { case challengeSignature = "challenge_signature", transport }
}

public struct SessionGrant: Codable, Sendable {
    public let ssidOrService: String
    public let psk: String
    public let ip: String
    public let token: String
    public let ttlSeconds: UInt32
    enum CodingKeys: String, CodingKey {
        case ssidOrService = "ssid_or_service", psk, ip, token, ttlSeconds = "ttl_seconds"
    }
}

// MARK: - Transfers

public enum TransferKind: String, Codable, Sendable { case offload, load }
public enum TransferMode: String, Codable, Sendable { case batch, trickle }

public struct TransferItem: Codable, Sendable, Equatable {
    public let id: String
    public let sizeBytes: UInt64
    public let mime: String
    enum CodingKeys: String, CodingKey { case id, sizeBytes = "size_bytes", mime }
}

public struct TransferRequest: Codable, Sendable {
    public let kind: TransferKind
    public let mode: TransferMode
    public let items: [TransferItem]
    public let temporaryPolicy: TemporaryPolicy?
    enum CodingKeys: String, CodingKey { case kind, mode, items, temporaryPolicy = "temporary_policy" }
}

public struct TransferProgress: Codable, Sendable, Equatable {
    public let itemsTotal: UInt32
    public let itemsDone: UInt32
    public let bytesTotal: UInt64
    public let bytesDone: UInt64
    public let etaSeconds: UInt32
    public let throughputBps: UInt64
    enum CodingKeys: String, CodingKey {
        case itemsTotal = "items_total", itemsDone = "items_done"
        case bytesTotal = "bytes_total", bytesDone = "bytes_done"
        case etaSeconds = "eta_seconds", throughputBps = "throughput_bps"
    }
}
